### DISCUSSION TITLE
3.13.6 breaks petal dropdown menus

### DISCUSSION THREAD

**Discussion by atomkirk** [alpinejs/alpine#4075] [7. Bugs (archived)]
I don't have many details, but I can confirm 3.13.6 breaks these dropdown menus https://petal.build/components/dropdowns

https://github.com/petalframework/petal_components/blob/main/lib/petal_components/dropdown.ex

I had my tag not pinned (mistake)  

```
<script defer src="https://unpkg.com/alpinejs@3.x.x/dist/cdn.min.js"></script>
```

and all the dropdowns on my site just stopped working. I pinned to 3.13.5 and they work again.

**Comment by ekwoka** [comment]
The seem to work on that site only after navigating to another page and back.

To me it looks like they are designed poorly and have two `click outside` listeners on different elements...so every click is a Click outside setting open to false...

**Reply by mitkins** [reply]
I think it might be related to this discussion - https://github.com/alpinejs/alpine/discussions/4010. 

When you visit a page for the first time in LiveView - the page is mounted twice:

* Once for the initial render (HTTP request); and 
* Once again after the JavaScript client connects to the server (usually over a web socket)

In LiveView I'm assuming there are mutations on the DOM between mounts (even though the contents itself hasn't changed). The following code is supposed to ensure that Alpine retains state (after mutation):

```JavaScript
let liveSocket = new LiveSocket("/live", Socket, {
  ...
  dom: {
    onBeforeElUpdated(from, to) {
      if (from._x_dataStack) {
        window.Alpine.clone(from, to);
      }
    },
  },
  ...
});
```

If this code fails (for some reason), then two events will be registered for the same element. 

I think perhaps that this is the case - and explains your (correct) assessement. A single element is being assigned two `click outside` listeners. When you navigate to another LiveView and come back, there will be only one event listener.

This is happening to me with the above code (in the past it wasn't a problem). When I used the code in the referenced discussion:

```JavaScript
let liveSocket = new LiveSocket("/live", Socket, {
  ...
  dom: {
    onBeforeElUpdated(from, to) {
      if (from && from._x_isInit) {
        window.Alpine.cloneNode(from, to);
      }
    },
  },
  ...
});
```

It worked!

So does that mean that `Alpine.clone` has changed across releases?

P.S. I'm not sure if the re-written method is the recommended way to do it. But the PR was merged

**Reply by mitkins** [reply]
Though it seems that looking through `main`, that `_x_isInit` isn't a thing. So would this be correct?

```JavaScript
if (from._x_dataStack) {
  window.Alpine.cloneNode(from, to);
}
```

**Reply by mitkins** [reply]
I've been doing more testing. Although the above code helped with the dropdown component - other parts of https://petal.build/ didn't work like they once had (e.g. parts of the main menu don't automatically collapse when they should on second mount).

I thought I might test specific versions of Alpine.js so I could identify the last known version that worked. Using the following code:

```JavaScript
let liveSocket = new LiveSocket("/live", Socket, {
  ...
  dom: {
    onBeforeElUpdated(from, to) {
      if (from._x_dataStack) {
        window.Alpine.clone(from, to);
      }
    },
  },
  ...
});
```

I tested the following versions:

* 3.13.5 to 3.14.3 worked as expected
* 3.14.4 to 3.14.7 did not

I did try `cloneNode` for 3.14.4 to 3.14.7, but I experienced the issues mentioned above.

**Comment by FeeJai** [comment]
Thank you @atomkirk. This post just solved my issue

Edit: I created a ticket at Petal Components: https://github.com/petalframework/petal_components/issues/379

**Reply by Matsa59** [reply]
It's not related to petal. The problem comes from the interaction between LV and AlpineJS.

- clone works for the websocket mount
- clodeNode almost works with some bug for stream update event from LV

Each community (alpinejs and phoenix) doesn't want to fix that and send the problem to each other. They both doesn't handle properly change on the dom but no one want to change things because it could break other part of the framework.

**Reply by Matsa59** [reply]
soooo after updating to phoenix LV 0.20.17 (yeah we're not on the 1.x.x for the moment) we had a bug based on a previous workarround.

TLDR: new workarround

Phoenix LV: 0.20.17
AlpineJS: 3.14.8

```diff
const liveSocket = new LiveSocket('/live', Socket, {
  uploaders: uploaders,
  dom: {
+   onNodeAdded(el) {
+     if (el.nodeType == 1 && el._x_marker) {
+       el._x_marker = undefined;
+       window.Alpine.initTree(el);
+     }
+   },
    onBeforeElUpdated(from, to) {
      if(from._x_dataStack){
        window.Alpine.clone(from, to);
      }
    }
  },
  hooks: Hooks,
  params: { _csrf_token: csrfToken },
});
```

---

Phoenix LV had the GREAT idea to "restore" element on the dom when there is a reset in the stream. What a great idea ... 
So alpine will consider the node as deleted (normal behaviour IMO). BUT Because LV restore it, you had a node with its state reseted ... If this happen, we'll have a onNodeAdded event trigger, so just "reset" _x_marker so from alpine POV it's like this element has never been initialized. And so it re-init it.

I don't know if it works for every case but it seems to work everywhere
